### ISSUE TITLE
incompatibility with Audaki Cart Engine

### ISSUE THREAD

**Issue by peow** [iluha168/MCTheCopperRail#2]
Audaki Cart Engine (https://modrinth.com/mod/audaki-cart-engine) basically makes the minecart faster but without breaking redstone circuits, but when using it together with the Copper Rail mod, the cart ignores the effect of the Copper Rail and detects it as just a normal rail, and this problem **only** happens when the Player is driving the cart, when the cart is empty everything works correctly

I showed this problem to the creator of Audaki Cart Engine and he told me this:

              Heya, I just checked it and the mod in question is also overriding the function moveOnRail: https://github.com/iluha168/MCTheCopperRail/blob/main/src/main/java/com/thecopperrail/mixin/AbstractMinecartMixin.java#L23C20-L23C30 ; I don’t know how to properly fix this, but I will accept pull requests as long as they don’t negatively impact performance for people who don’t use this mod.

_by @audaki in https://github.com/audaki/minecraft-cart-engine/issues/26#issuecomment-2028455708_

-------------

if it is possible to fix this I would appreciate it if there was an update to fabric **1.20.1**

**Comment by iluha168** [comment]
```pseudocode
minecartMovementTick {
    if minecart is being ridden {
        vanilla logic but faster movement
    } else {
        vanilla logic with copper rails
    }
}
```

**Comment by iluha168** [comment]
The incompatibility comes from ACE mod. Authors use mixins irresponsibly, if that makes sense.

Let me explain it to you. Imagine this pseudocode is vanilla code for minecart movement:
```pseudocode
minecartMovementTick {
    vanilla logic
}
```
What mods should do (to be as most compatible with other mods as possible) is modify existing vanilla logic bit by bit:
```pseudocode
minecartMovementTick {
    V
    A  <- (mod 1 affect this)
    N
    I
    L  <- (mod 2 affects that)
    L
    A
}
```
Let us call this slightly modified logic modded logic
```pseudocode
minecartMovementTick {
        modded logic
}
```
The problem is, the authors of ACE mod copied the entirety of decompiled vanilla logic, modified it, and inject back the copy:
```pseudocode
minecartMovementTick {
    if minecart is rideable and is being ridden:
        vanilla logic but faster movement
    else:
        modded logic
}
```
This results in vanilla-rails-only logic when a minecart is being ridden. As you can see, most (all?) modded rails will not work because of this. You can prove this by installing more rail mods and noticing none of them work when a minecart is being ridden.
The copper rail acts as a golden rail instead because it inherits from PoweredRailBlock.
**TL;DR** This is not my problem :P ACE mod needs fixing instead.

Regarding your second question, the mod already works for all 1.20.x versions as far as i tested. If it does not, you can always make another issue (make sure to specify the version then).

**Comment by peow** [comment]
yes, your mod already works in 1.20.1, I was referring to a possible update with compatibility with the Audaki Cart Engine

**Comment by audaki** [comment]
@iluha168 Before you throw punches like this you should have a closer look at code you intend to criticize, otherwise it might turn out to be a tad embarrassing for yourself in the end. 

To use your example:

moveOnRail {
    V  <- (ACE affects this)
    A  <- (ACE affects this)
    N  <- (ACE affects this)
    I  <- (ACE affects this)
    L  <- (ACE affects this)
    L  <- (ACE affects this)
    A  <- (ACE affects this)
}

Hope this helps you to understand.

**Comment by iluha168** [comment]
@audaki [That is wrong.](https://github.com/audaki/minecraft-cart-engine/blob/fabric-mc1.20/src/main/java/audaki/cart_engine/mixin/AbstractMinecartEntityMixin.java#L69)

Decompiled by FernFlower AbstractMinecartEntity.class after applying mixins from ACE-2.0.6 and TCR 0.9.1
```java
   public void moveOnRail(BlockPos pos, BlockState state) {
      CallbackInfo callbackInfo55 = new CallbackInfo("moveOnRail", true);
      this.handler$zcj000$audaki_cart_engine$moveOnRailOverwrite(pos, state, callbackInfo55);
      if (!callbackInfo55.isCancelled()) {
         this.onLanding();
         double d = this.getX();
         double e = this.getY();
         double f = this.getZ();
         Vec3d vec3d = this.snapPositionToRail(d, e, f);
         e = (double)pos.getY();
         boolean bl = false;
         boolean bl2 = false;
         Block injectorAllocatedLocal57 = Blocks.POWERED_RAIL;
         if (this.redirect$zig000$thecopperrail$injectedPoweredRailCheck(state, injectorAllocatedLocal57)) {
            bl = (Boolean)state.get(PoweredRailBlock.POWERED);
            bl2 = !bl;
         }

         double g = 0.0078125;
         if (this.isTouchingWater()) {
            g *= 0.2;
         }

         Vec3d vec3d2 = this.getVelocity();
         RailShape railShape = (RailShape)state.get(((AbstractRailBlock)state.getBlock()).getShapeProperty());
         switch (1.field_7682[railShape.ordinal()]) {
            case 1:
               this.setVelocity(vec3d2.add(-g, 0.0, 0.0));
               ++e;
               break;
            case 2:
               this.setVelocity(vec3d2.add(g, 0.0, 0.0));
               ++e;
               break;
            case 3:
               this.setVelocity(vec3d2.add(0.0, 0.0, g));
               ++e;
               break;
            case 4:
               this.setVelocity(vec3d2.add(0.0, 0.0, -g));
               ++e;
         }

         vec3d2 = this.getVelocity();
         Pair<Vec3i, Vec3i> pair = getAdjacentRailPositionsByShape(railShape);
         Vec3i vec3i = (Vec3i)pair.getFirst();
         Vec3i vec3i2 = (Vec3i)pair.getSecond();
         double h = (double)(vec3i2.getX() - vec3i.getX());
         double i = (double)(vec3i2.getZ() - vec3i.getZ());
         double j = Math.sqrt(h * h + i * i);
         double k = vec3d2.x * h + vec3d2.z * i;
         if (k < 0.0) {
            h = -h;
            i = -i;
         }

         double l = Math.min(2.0, vec3d2.horizontalLength());
         vec3d2 = new Vec3d(l * h / j, vec3d2.y, l * i / j);
         this.setVelocity(vec3d2);
         Entity entity = this.getFirstPassenger();
         if (entity instanceof PlayerEntity) {
            Vec3d vec3d3 = entity.getVelocity();
            double m = vec3d3.horizontalLengthSquared();
            double n = this.getVelocity().horizontalLengthSquared();
            if (m > 1.0E-4 && n < 0.01) {
               this.setVelocity(this.getVelocity().add(vec3d3.x * 0.1, 0.0, vec3d3.z * 0.1));
               bl2 = false;
            }
         }

         double o;
         if (bl2) {
            o = this.getVelocity().horizontalLength();
            if (o < 0.03) {
               this.setVelocity(Vec3d.ZERO);
            } else {
               this.setVelocity(this.getVelocity().multiply(0.5, 0.0, 0.5));
            }
         }

         o = (double)pos.getX() + 0.5 + (double)vec3i.getX() * 0.5;
         double p = (double)pos.getZ() + 0.5 + (double)vec3i.getZ() * 0.5;
         double q = (double)pos.getX() + 0.5 + (double)vec3i2.getX() * 0.5;
         double r = (double)pos.getZ() + 0.5 + (double)vec3i2.getZ() * 0.5;
         h = q - o;
         i = r - p;
         double s;
         double t;
         double u;
         if (h == 0.0) {
            s = f - (double)pos.getZ();
         } else if (i == 0.0) {
            s = d - (double)pos.getX();
         } else {
            t = d - o;
            u = f - p;
            s = (t * h + u * i) * 2.0;
         }

         d = o + h * s;
         f = p + i * s;
         this.setPosition(d, e, f);
         t = this.hasPassengers() ? 0.75 : 1.0;
         u = this.getMaxSpeed();
         vec3d2 = this.getVelocity();
         this.move(MovementType.SELF, new Vec3d(MathHelper.clamp(t * vec3d2.x, -u, u), 0.0, MathHelper.clamp(t * vec3d2.z, -u, u)));
         if (vec3i.getY() != 0 && MathHelper.floor(this.getX()) - pos.getX() == vec3i.getX() && MathHelper.floor(this.getZ()) - pos.getZ() == vec3i.getZ()) {
            this.setPosition(this.getX(), this.getY() + (double)vec3i.getY(), this.getZ());
         } else if (vec3i2.getY() != 0 && MathHelper.floor(this.getX()) - pos.getX() == vec3i2.getX() && MathHelper.floor(this.getZ()) - pos.getZ() == vec3i2.getZ()) {
            this.setPosition(this.getX(), this.getY() + (double)vec3i2.getY(), this.getZ());
         }

         this.applySlowdown();
         Vec3d vec3d4 = this.snapPositionToRail(this.getX(), this.getY(), this.getZ());
         Vec3d vec3d5;
         double w;
         if (vec3d4 != null && vec3d != null) {
            double v = (vec3d.y - vec3d4.y) * 0.05;
            vec3d5 = this.getVelocity();
            w = vec3d5.horizontalLength();
            if (w > 0.0) {
               this.setVelocity(vec3d5.multiply((w + v) / w, 1.0, (w + v) / w));
            }

            this.setPosition(this.getX(), vec3d4.y, this.getZ());
         }

         int x = MathHelper.floor(this.getX());
         int y = MathHelper.floor(this.getZ());
         if (x != pos.getX() || y != pos.getZ()) {
            vec3d5 = this.getVelocity();
            w = vec3d5.horizontalLength();
            this.setVelocity(w * (double)(x - pos.getX()), vec3d5.y, w * (double)(y - pos.getZ()));
         }

         if (bl) {
            Vec3d var10000 = this.getVelocity();
            CallbackInfo callbackInfo58 = new CallbackInfo("moveOnRail", true);
            this.handler$zig000$thecopperrail$injectedCopperRailCallback(pos, state, callbackInfo58);
            if (callbackInfo58.isCancelled()) {
               return;
            }

            vec3d5 = var10000;
            w = vec3d5.horizontalLength();
            if (w > 0.01) {
               double z = 0.06;
               this.setVelocity(vec3d5.add(vec3d5.x / w * 0.06, 0.0, vec3d5.z / w * 0.06));
            } else {
               Vec3d vec3d6 = this.getVelocity();
               double aa = vec3d6.x;
               double ab = vec3d6.z;
               if (railShape == RailShape.EAST_WEST) {
                  if (this.willHitBlockAt(pos.west())) {
                     aa = 0.02;
                  } else if (this.willHitBlockAt(pos.east())) {
                     aa = -0.02;
                  }
               } else {
                  if (railShape != RailShape.NORTH_SOUTH) {
                     return;
                  }

                  if (this.willHitBlockAt(pos.north())) {
                     ab = 0.02;
                  } else if (this.willHitBlockAt(pos.south())) {
                     ab = -0.02;
                  }
               }

               this.setVelocity(aa, vec3d6.y, ab);
            }
         }

      }
   }
```

**Comment by peow** [comment]
I think that a respectful conversation is the most likely to generate good results, because at the end of the day the incompatibility between the Copper Rail and the Audaki Cart Engine is not a problem in the Copper Rail or in the Audaki Cart Engine, but rather a problem in the end user experience, who will have to give up a great mod for another

**Comment by Blockmeska** [comment]
Hello, @iluha168. I think you can fix this by changing the priority of your mixin to higher than the other mod (to be injected first). While the other mod completely replaces the moveOnRail method with a new one, if you manage to execute your code first, it might just work fine. Currently, the other mod seems to have a default priority for its mixin.

**Comment by iluha168** [comment]
@Blockmeska you can fix this by rewriting ACE. ACE creates a copy of the vanilla function, meaning other mods would have to inject the same code twice into 2 different locations with 2 different priorities in order to be "compatible".
I could make my own yet another minecart boosting mod if you guys really want this? :thinking:

**Comment by peow** [comment]
would be cool, how do you imagine it?
doing some tests, I concluded that this mod here is currently the best for Fabric 1.20.1 https://modrinth.com/mod/fast-minecart

other mods have things that I don't like, such as the carts loading chunks, and the Player's animation on curves being strange, where it looks like the Player's body leaves the cart and returns

The problem with this mod is about redstone, the carts are so fast that I don't think it's possible to make systems with them
